File: sdk/java/src/test/java/io/cordum/cap/ProgressTest.java
package io.cordum.cap;

import io.cordum.cap.agent.v1.BusPacket;
import io.cordum.cap.agent.v1.JobCancel;
import io.cordum.cap.agent.v1.JobProgress;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ProgressTest {

    @Test
    void progressPayloadProducesValidPacket() throws Exception {
        byte[] bytes = Progress.progressPayload("sender-1", "job-42", "step-3", 75, "processing");
        assertTrue(bytes.length > 0);

        BusPacket packet = BusPacket.parseFrom(bytes);
        assertEquals("sender-1", packet.getSenderId());
        assertEquals(1, packet.getProtocolVersion());
        assertTrue(packet.hasJobProgress());

        JobProgress p = packet.getJobProgress();
        assertEquals("job-42", p.getJobId());
        assertEquals("step-3", p.getStepId());
        assertEquals(75, p.getPercent());
        assertEquals("processing", p.getMessage());
    }

    @Test
    void progressPayloadZeroPercent() throws Exception {
        byte[] bytes = Progress.progressPayload("s", "j", "st", 0, "");
        BusPacket packet = BusPacket.parseFrom(bytes);
        assertTrue(packet.hasJobProgress());
        assertEquals(0, packet.getJobProgress().getPercent());
    }

    @Test
    void cancelPayloadProducesValidPacket() throws Exception {
        byte[] bytes = Progress.cancelPayload("sender-1", "job-99", "timeout", "admin");
        assertTrue(bytes.length > 0);

        BusPacket packet = BusPacket.parseFrom(bytes);
        assertEquals("sender-1", packet.getSenderId());
        assertTrue(packet.hasJobCancel());

        JobCancel c = packet.getJobCancel();
        assertEquals("job-99", c.getJobId());
        assertEquals("timeout", c.getReason());
        assertEquals("admin", c.getRequestedBy());
    }

    @Test
    void progressPayloadDeterministic() {
        byte[] a = Progress.progressPayload("s", "j", "st", 50, "half");
        byte[] b = Progress.progressPayload("s", "j", "st", 50, "half");
        // Note: timestamps differ, so we can't assertArrayEquals, but both should be valid
        assertTrue(a.length > 0);
        assertTrue(b.length > 0);
    }

    @Test
    void cancelPayloadDeterministic() {
        byte[] a = Progress.cancelPayload("s", "j", "r", "u");
        byte[] b = Progress.cancelPayload("s", "j", "r", "u");
        assertTrue(a.length > 0);
        assertTrue(b.length > 0);
    }
}

File: sdk/php/tests/ProgressTest.php
<?php

declare(strict_types=1);

namespace Cordum\Cap\Tests;

use Cordum\Agent\V1\BusPacket;
use Cordum\Cap\ProgressHelper;
use PHPUnit\Framework\TestCase;

class ProgressTest extends TestCase
{
    #[\PHPUnit\Framework\Attributes\Test]
    public function progressPayloadValid(): void
    {
        $bytes = ProgressHelper::progressPayload('s-1', 'j-42', 'step-3', 75, 'processing');
        $this->assertNotEmpty($bytes);

        $packet = new BusPacket();
        $packet->mergeFromString($bytes);
        $this->assertSame('s-1', $packet->getSenderId());

        $p = $packet->getJobProgress();
        $this->assertNotNull($p);
        $this->assertSame('j-42', $p->getJobId());
        $this->assertSame('step-3', $p->getStepId());
        $this->assertSame(75, $p->getPercent());
        $this->assertSame('processing', $p->getMessage());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function cancelPayloadValid(): void
    {
        $bytes = ProgressHelper::cancelPayload('s-1', 'j-99', 'timeout', 'admin');
        $this->assertNotEmpty($bytes);

        $packet = new BusPacket();
        $packet->mergeFromString($bytes);

        $c = $packet->getJobCancel();
        $this->assertNotNull($c);
        $this->assertSame('j-99', $c->getJobId());
        $this->assertSame('timeout', $c->getReason());
        $this->assertSame('admin', $c->getRequestedBy());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function progressPayloadZeroPercent(): void
    {
        $bytes = ProgressHelper::progressPayload('s', 'j', 'st', 0, '');
        $packet = new BusPacket();
        $packet->mergeFromString($bytes);
        $this->assertSame(0, $packet->getJobProgress()->getPercent());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function progressPayload100Percent(): void
    {
        $bytes = ProgressHelper::progressPayload('s', 'j', 'st', 100, 'done');
        $packet = new BusPacket();
        $packet->mergeFromString($bytes);
        $this->assertSame(100, $packet->getJobProgress()->getPercent());
        $this->assertSame('done', $packet->getJobProgress()->getMessage());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function progressPayloadHasTimestamp(): void
    {
        $bytes = ProgressHelper::progressPayload('s', 'j', 'st', 50, 'half');
        $packet = new BusPacket();
        $packet->mergeFromString($bytes);
        $this->assertNotNull($packet->getCreatedAt());
        $this->assertGreaterThan(0, $packet->getCreatedAt()->getSeconds());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function cancelPayloadHasTimestamp(): void
    {
        $bytes = ProgressHelper::cancelPayload('s', 'j', 'r', 'u');
        $packet = new BusPacket();
        $packet->mergeFromString($bytes);
        $this->assertNotNull($packet->getCreatedAt());
    }
}
